### PR TITLE
[NPU][dLLM] Stop materializing fp32 [tokens, vocab] logits in the denoise step

### PR DESCRIPTION
**Pull request by wenxuewuhd**
## Motivation

A dLLM denoise step needs exactly two things per position: the argmax token, and the softmax probability of that same token — the confidence it compares against the unmask threshold. Both are reductions over the vocab.

The logits path hands the algorithms a full fp32 `[tokens, vocab]` tensor to compute them from, and both algorithms then build a second full-width tensor for the softmax. At LLaDA2's vocab of 157184 that is **2.1 GiB per copy at 112 blocks** — this, not the KV pool, is what caps the dLLM batch size, and it is pure waste: nothing downstream needs the full matrix.

## Modifications

Add `argmax_and_softmax_prob()` in `dllm/algorithm/base.py` and route both denoise algorithms through it. It has three implementations of the same quantity and picks by what the caller already paid for:

1. **`sgl_kernel_npu.sample.argmax_softmax_prob`**, when that kernel is installed. It reads each logit once, carries `(max, argmax, sumexp)` in registers via the online-softmax recurrence, and returns `(argmax, 1/sumexp)` — the argmax token's logit *is* the running max, so its shifted exponent is exactly 1 and its probability is the reciprocal of the denominator. Optional dependency; without it the branches below give the same answer.
2. **`argmax` + `softmax` + `gather`** when the logits are already fp32 — the pre-existing formulation, bitwise unchanged.
3. **Otherwise** a max reduction plus a vocab-chunked exp-sum, whose transients are `[tokens, vocab_chunk]` rather than `[tokens, vocab]`.

On NPU, `LogitsProcessor._get_dllm_logits` then keeps the lm_head dtype instead of copying into the fp32 logits buffer, **once the batch is wide enough**. That skipped copy is what lifts the memory ceiling, and it does not depend on the kernel: without it the logits land on branch 3, whose transients are one vocab strip rather than the full matrix. The kernel makes that path faster; it is not what makes it fit.

The width condition is not a hedge — it is where the two costs actually sit. The reduction over low-precision logits is launch-bound (a 157k vocab is ~10 chunks), so at one denoise block it is 9x slower than the fp32 formulation, while at 128 blocks it is the only thing that fits in memory. Below the threshold there is no memory pressure to trade against, so the fp32 path is kept and nothing changes. Default 512 rows on Ascend, where the crossover was measured; unset on backends where it has not been. See Configuration.

The one place that still needs fp32 is the repetition-penalty branch, which rewrites logits before the argmax; it takes an explicit `.float()` copy, a no-op when the logits are already fp32.

### Blast radius

| File | Scope |
| --- | --- |
| `dllm/algorithm/base.py`, `joint_threshold.py`, `low_confidence.py` | dLLM only — these files have no non-dLLM callers. |
| `layers/logits_processor.py` | **Shared file, not env-gated.** See below. |

`_get_logits` gains `keep_lm_head_dtype: bool = False`. Both pre-existing call sites omit it and therefore take the identical `_copy_logits_to_buffer` branch — the added code is unreachable for them.

`_copy_logits_to_buffer` does two things, and the new branch reproduces only one. The padding trim (`logits[:, :vocab_size]`) is carried over. Its other job — publishing into `next_token_logits_buffer`, the graph-captured static buffer — has no counterpart on this path: `_get_dllm_logits` returns `full_logits` with `next_token_logits` left `None`, so nothing reads that buffer for a dLLM forward.

The one caller that passes the flag is `_get_dllm_logits`, itself only reachable from `LogitsProcessor.forward` when `return_full_logits=True`, which in-tree is llada2, sdar and sdar_moe only. And it passes a flag that is false on any backend without a measured threshold, so on every other backend the function is byte-identical to today — as it is on NPU below the threshold.

The dtype branch in `argmax_and_softmax_prob` exists for the same reason. Without it, non-NPU dLLM would take the chunked reduction on logits that are still fp32 — paying for both the chunking and the fp32 copy. Measured at 16.8 ms against 9.4 ms for the code it replaces, i.e. **1.8x slower on the default GPU path with no flag to turn it off**. With the branch, non-NPU dLLM is bitwise what it is today; the unit test asserts that with `torch.equal`.

### Configuration

One new environment variable, `SGLANG_DLLM_KEEP_LOGITS_DTYPE_MIN_ROWS` (`EnvInt(None)`), the denoise row count at which the logits stay in lm_head dtype. It follows the same shape as `SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS` in #33047 — a measured per-backend default, an integer override, and `-1` to disable — but it is a separate knob: the two gate different quantities and their measured crossovers differ by two orders of magnitude, so one value cannot serve both.

| Value | Behavior | When to use |
| --- | --- | --- |
| unset (default) | Ascend: keep lm_head dtype at 512+ rows. Other backends: always fp32. | Normal serving, both latency and throughput. |
| `N` (>= 0) | Keep it at `N`+ rows. | Enabling it on a backend after measuring its crossover, or retuning Ascend. |
| `-1` | Never. | Production A/B and quick rollback. |

The resolved default is logged once at INFO.

## Accuracy Tests

The fused kernel is not bit-identical to the fp32 reference (it accumulates the same sum in a different order), so it was checked directly on Ascend 910B3 at vocab 157184, bf16 logits:

| Quantity | Result |
| --- | --- |
| argmax id | identical on every row |
| softmax probability | max relative error 3.5e-7 |
| decisions at a 0.5 unmask threshold | no flips |

Branches 2 and 3 need no such check: branch 2 *is* the pre-existing expression, and branch 3 differs from it only by chunking a sum.

End to end, LLaDA2-mini gsm8k (zero-shot, full 1319 samples): accuracy unchanged.

## Speed Tests and Profiling

Measured on Ascend 910B3, LLaDA2-mini logits (vocab 157184), per denoise step, all three paths against each other:

| rows | blocks | before (fp32) | after, no kernel | after, with kernel |
| ---: | ---: | ---: | ---: | ---: |
| 32 | 1 | 0.11 ms | 1.03 ms | 0.19 ms |
| 128 | 4 | 0.38 ms | 1.01 ms | 0.20 ms |
| 512 | 16 | 1.67 ms | 1.37 ms | 0.62 ms |
| 1024 | 32 | 3.26 ms | 2.52 ms | 1.28 ms |
| 2048 | 64 | 6.34 ms | 4.90 ms | 2.54 ms |
| 4096 | 128 | 12.40 ms | 13.41 ms | 5.02 ms |

Two things this table decides. First, the threshold: without the kernel the crossover is 512 rows, and below it the loss is steep rather than marginal — 9x at one block, because a 157k vocab chunks into ~10 launches while the fp32 formulation is three fused ops. Second, that the threshold has to exist at all rather than gating on kernel availability: even with the kernel, one block is 1.6x slower than the fp32 path.

At and above the threshold the kernel is a flat ~2.5x, and the no-kernel path is 1.2-1.3x up to 64 blocks, turning into an 8% loss at 128. That last cell is the trade the memory result is worth: 128 blocks is where the fp32 copy stops fitting.

The memory result does not wait for the kernel. Above the threshold the fp32 `[tokens, vocab]` copy is gone on NPU either way, which is what put a ceiling on the batch size at 112+ blocks.

## Unit Tests

`test/registered/unit/dllm/test_argmax_softmax_prob.py` (CPU-only, no server, `base-a-test-cpu`), 5 cases:

- the fp32 branch is bitwise the pre-existing `argmax` + `softmax` + `gather` formulation — this assertion *is* the "non-NPU unchanged" claim above, not a restatement of it;
- the chunked branch's `1/sumexp` identity, and its running-max rescale across a chunk boundary;
- lm_head-dtype input allocates nothing wider than one vocab strip — the memory claim, asserted with the fused kernel patched out, which is the configuration it has to hold in.

Mutation-checked: dropping the `- row_max` shift fails 3 of the 5.

The row threshold has no unit test: it is one comparison against a resolved constant, and the thing worth guarding — that the code below it is untouched — is already guaranteed structurally, since `keep_lm_head_dtype=False` reaches the identical pre-existing branch.

The kernel itself is tested in its own repository, where it can run on the hardware it targets.

## Dependency note

The fused path imports `sgl_kernel_npu.sample.argmax_softmax_prob` (sgl-project/sgl-kernel-npu#644). The import is guarded and every branch of `argmax_and_softmax_prob` returns the same values, so this PR is complete and correct without it.

The memory result does not depend on it either — branch 3 reduces in vocab strips, so the ceiling lifts on any NPU as soon as this PR lands.

What the kernel changes is speed above the threshold: 5.0 ms against 13.4 ms at 128 blocks. Without it that path is 8% slower than the fp32 formulation it replaces, which is the deliberate trade — the OOM is a hard failure at 112+ blocks, the 8% is a soft cost on a path that still runs, and it disappears on upgrade. Below the threshold the kernel is irrelevant either way, since the fp32 path is kept.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35508977487](https://github.com/sgl-project/sglang/actions/runs/35508977487)
Latest PR Test (Extra): :x: [Run #35508977363](https://github.com/sgl-project/sglang/actions/runs/35508977363)
Latest PR Test (AMD ROCm 10): :x: [Run #35508977494](https://github.com/sgl-project/sglang/actions/runs/35508977494)

### CODE REVIEW

File: python/sglang/srt/dllm/algorithm/base.py [hunk 1]
```diff
@@ -23,6 +32,60 @@
...
+        ).squeeze(-1)
+        return argmax_ids, prob
+
+    row_max, argmax_ids = torch.max(full_logits_2d, dim=-1)
+    row_max_col = row_max.unsqueeze(1)
+    denom: Optional[torch.Tensor] = None
+    for start in range(0, full_logits_2d.shape[1], vocab_chunk):
+        shifted = full_logits_2d[:, start : start + vocab_chunk] - row_max_col
+        shifted.exp_()
+        part = shifted.sum(dim=-1, dtype=torch.float32)
```
Comment (rwang5203) [inline python/sglang/srt/dllm/algorithm/base.py]: BF16 exponentiation rounds small deltas before the fp32 sum, changing strict-threshold decisions.

Exponentiate each bounded chunk in fp32 and test the `0.5` boundary.

File: test/registered/unit/dllm/test_argmax_softmax_prob.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        def record(t, *a, **kw):
+            widths.append(t.shape[-1] if t.dim() == 2 else 0)
+            return real_empty(t, *a, **kw)
+
+        with patch("torch.empty_like", record):
+            ids, prob = argmax_and_softmax_prob(logits, vocab_chunk=chunk)
+
+        self.assertEqual(prob.shape, (rows,))
+        self.assertTrue(all(w <= chunk for w in widths), widths)
```
Comment (rwang5203) [inline test/registered/unit/dllm/test_argmax_softmax_prob.py]: This allocation assertion is vacuous because `torch.empty_like` is never called.

Measure casts or peak memory and fail on a full-width fp32 copy.

**Comment by rwang5203** [comment]
Runtime [CI](https://github.com/sgl-project/sglang/actions/runs/30679164721/job/91312554904) skipped CPU, CUDA, NPU, and fused dispatch.

Run registered CPU/CUDA tests and NPU fp32/bf16 integration with [#644](https://github.com/sgl-project/sgl-kernel-npu/commit/b872afda5179ffb455736297bd6bdbbcf8a200d4).

File: python/sglang/srt/dllm/algorithm/base.py (hunk 1 above)
Comment (wenxuewuhd) [reply python/sglang/srt/dllm/algorithm/base.py]: Done. With the top two logits nearly tied (prob 0.50002..0.51251), bf16 exp puts 20/512 rows on the wrong side of 0.5. Each chunk is now cast to fp32 before the subtract and exp. Relative error at [512, 157184] drops 3.9e-04 → 7.1e-05, argmax unchanged. Boundary case added; reverting to bf16 exp turns it red.

The fp32 transient is still `[rows, vocab_chunk]`. At [512, 157184], chunk 16384: peak 16 MiB → 32 MiB, versus 307 MiB for a full-width copy, and zero full-width fp32 allocations either way. At bs=128 that's +128 MiB against 2.4 GiB avoided.

File: test/registered/unit/dllm/test_argmax_softmax_prob.py (hunk 2 above)
Comment (wenxuewuhd) [reply test/registered/unit/dllm/test_argmax_softmax_prob.py]: Confirmed — `torch.empty_like` is called 0 times there, so it asserted over an empty list.

Now uses `TorchDispatchMode`, so it no longer depends on which allocating API the implementation calls: bounds every fp32 2-D tensor to `vocab_chunk`, and asserts the list is non-empty so it can't go vacuous again. A full-width fp32 copy turns it red.

**Comment by wenxuewuhd** [comment]
> Runtime [CI](https://github.com/sgl-project/sglang/actions/runs/30679164721/job/91312554904) skipped CPU, CUDA, NPU, and fused dispatch.
> 
> Run registered CPU/CUDA tests and NPU fp32/bf16 integration with [#644](https://github.com/sgl-project/sgl-kernel-npu/commit/b872afda5179ffb455736297bd6bdbbcf8a200d4).

CI didn't run: `call-gate / pr-gate` fails on the run-ci label step, so everything downstream is skipped. The tests are registered — they'll run once the label is on, and fused dispatch comes with them: `npu_ci_install_dependency.sh` pins `SGLANG_KERNEL_NPU_TAG=2026.8.10`, which is the release carrying #644, so the Ascend dLLM job exercises the fused path rather than the portable one.

Ran them locally meanwhile: registered CPU 6 passed; registered CUDA `test_dllm_fdfo_joint_threshold` passed on H20; Ascend dLLM e2e with #644 2 passed, gsm8k 0.93; #644's own kernel tests 12 passed. CUDA takes the fp32 branch by default, so I also forced the reduction path on with `SGLANG_DLLM_KEEP_LOGITS_DTYPE_MIN_ROWS=1` — gsm8k 0.830, 0 invalid.

File: python/sglang/srt/dllm/algorithm/base.py [hunk 3]
```diff
@@ -23,6 +32,59 @@
...
+      fp32 cast internally.
+
+    The argmax is identical across all three; the probability differs only by
+    exp's precision in the input dtype.
+    """
+    if _argmax_softmax_prob_fused is not None:
+        # Single-pass fused kernel: one read of the logits vs the 5-6 below.
+        return _argmax_softmax_prob_fused(full_logits_2d)
+
+    if full_logits_2d.dtype == torch.float32:
```
Comment (rwang5203) [inline python/sglang/srt/dllm/algorithm/base.py]: The fused branch precedes the fp32 guard, so installing the optional kernel bypasses the preserved latency and numeric path for below-threshold batches.

Gate fused dispatch on retained lm-head dtype and add an installed-kernel fp32 regression.

Comment (wenxuewuhd) [reply python/sglang/srt/dllm/algorithm/base.py]: Fixed in 6262361 — the fp32 check now precedes the fused-kernel check, so the below-threshold batches that `logits_processor` keeps in fp32 stay on the pre-existing path even with the kernel installed; the fused kernel only ever sees lm_head-dtype logits. Added a regression test that patches in a fused-kernel mock that raises if called and requires bitwise equality with the fp32 reference — it fails against the old order.
